jdaviz/app.py

```diff
@@ -1069,7 +1069,63 @@ def get_viewer_ids(self, prefix=None):
+            self, require_no_selected_data=False,
+            require_spectrum_viewer=False,
+            require_table_viewer=False,
+            require_flux_viewer=False,
+            require_image_viewer=False
+    ):
+        """
+        Return the viewer reference name of the first available viewer.
+        Optionally use ``require_no_selected_data`` to require that the
+        viewer has not yet loaded data, or e.g. ``require_spectrum_viewer``
+        to require that the viewer supports spectrum visualization.
+        """
+        from jdaviz.configs.specviz.plugins.viewers import SpecvizProfileView
+        from jdaviz.configs.cubeviz.plugins.viewers import CubevizProfileView, CubevizImageView
+        from jdaviz.configs.mosviz.plugins.viewers import (
+            MosvizProfileView, MosvizTableViewer, MosvizProfile2DView
+        )
+
+        spectral_viewers = (SpecvizProfileView, CubevizProfileView, MosvizProfileView)
+        table_viewers = (MosvizTableViewer, )
+        image_viewers = (MosvizProfile2DView, )
+        flux_viewers = (CubevizImageView, )
+
+        for vid in self._viewer_store:
+            viewer_item = self._viewer_item_by_id(vid)
+            if require_spectrum_viewer:
+                if isinstance(self._viewer_store[vid], spectral_viewers):
+                    if require_no_selected_data and not len(viewer_item['selected_data_items']):
+                        return viewer_item['reference']
+                    elif not require_no_selected_data:
+                        return viewer_item['reference']
+            elif require_table_viewer:
+                if isinstance(self._viewer_store[vid], table_viewers):
+                    if require_no_selected_data and not len(viewer_item['selected_data_items']):
+                        return viewer_item['reference']
+                    elif not require_no_selected_data:
+                        return viewer_item['reference']
+            elif require_image_viewer:
+                if isinstance(self._viewer_store[vid], image_viewers):
+                    if require_no_selected_data and not len(viewer_item['selected_data_items']):
+                        return viewer_item['reference']
+                    elif not require_no_selected_data:
+                        return viewer_item['reference']
+            elif require_flux_viewer:
+                if isinstance(self._viewer_store[vid], flux_viewers):
+                    if require_no_selected_data and not len(viewer_item['selected_data_items']):
+                        return viewer_item['reference']
+                    elif not require_no_selected_data:
+                        return viewer_item['reference']
+            else:
+                if require_no_selected_data and not len(viewer_item['selected_data_items']):
+                    return viewer_item['reference']
+                elif not require_no_selected_data:
+                    return viewer_item['reference']
```

Comment:
This repeated block of code is a bit ugly, but I wasn't sure how best to produce this behavior without redundancy. Any tips on how to make this concise?